opm/simulators/utils/HDF5File.cpp

```diff
@@ -0,0 +1,121 @@
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#include <config.h>
+#include <opm/simulators/utils/HDF5File.hpp>
+
+#include <filesystem>
+#include <stdexcept>
+
+namespace Opm {
+
+HDF5File::HDF5File(const std::string& fileName, OpenMode mode)
+{
+    bool exists = std::filesystem::exists(fileName);
+    if (mode == OpenMode::WRITE ||
+        (mode == OpenMode::APPEND && !exists)) {
+        m_file = H5Fcreate(fileName.c_str(),
+                           H5F_ACC_TRUNC,
+                           H5P_DEFAULT, H5P_DEFAULT);
+    } else {
+        m_file = H5Fopen(fileName.c_str(),
+                         mode == OpenMode::READ ? H5F_ACC_RDONLY : H5F_ACC_RDWR,
+                         H5P_DEFAULT);
+        if (m_file == H5I_INVALID_HID) {
+            throw std::runtime_error("HDF5File: Failed to open " + fileName);
+        }
+    }
+}
+
+HDF5File::~HDF5File()
+{
+    H5Fclose(m_file);
+}
+
+bool HDF5File::write(const std::string& group,
+                     const std::string& dset,
+                     const std::vector<char>& buffer)
+{
+    hid_t grp;
+    if (groupExists(m_file, group)) {
+        grp = H5Gopen2(m_file, group.c_str(), H5P_DEFAULT);
+    } else {
+        grp = H5Gcreate2(m_file, group.c_str(), 0, H5P_DEFAULT, H5P_DEFAULT);
+    }
+
+    hsize_t size = buffer.size();
+    hsize_t start = 0;
+
+    hid_t space = H5Screate_simple(1, &size, nullptr);
+    hid_t set = H5Dcreate2(grp, dset.c_str(), H5T_NATIVE_CHAR, space,
+                           H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+    if (set == H5I_INVALID_HID) {
+        throw std::runtime_error("HDF5File: Trying to write already existent dataset " + group + '/' + dset);
+    }
+
+    if (size > 0) {
+        hid_t filespace = H5Dget_space(set);
+        hsize_t stride = 1;
+        H5Sselect_hyperslab(filespace, H5S_SELECT_SET, &start, &stride, &size, nullptr);
+        hid_t memspace = H5Screate_simple(1, &size, nullptr);
+        H5Dwrite(set, H5T_NATIVE_CHAR, memspace, filespace, H5P_DEFAULT, buffer.data());
+        H5Sclose(memspace);
+        H5Sclose(filespace);
+    }
+    H5Dclose(set);
+    H5Sclose(space);
+    H5Gclose(grp);
+
+    return true;
+}
+
+bool HDF5File::read(const std::string& group,
+                    const std::string& dset,
+                    std::vector<char>& buffer)
+{
+    hid_t set = H5Dopen2(m_file, (group + "/"+ dset).c_str(), H5P_DEFAULT);
+    if (set == H5I_INVALID_HID) {
+        throw std::runtime_error("HDF5File: Trying to read nonexistent dataset " + group + '/' + dset);
+    }
+
+    hid_t space = H5Dget_space(set);
+    hsize_t size = H5Sget_simple_extent_npoints(space);
+    buffer.resize(size);
+    H5Dread(set, H5T_NATIVE_CHAR, H5S_ALL, H5S_ALL, H5P_DEFAULT, buffer.data());
+    H5Dclose(set);
+
+    return true;
```

Comment:
Why not make this function void? Alternatively, return false instead of throwing.

Also in write().

Reply:
no reason. copy-pasted from some other code where it was bool returns and no exceptions.. they are now void.